### PR TITLE
impl(rest): add HttpPayload class

### PR DESCRIPTION
**Pull request by scotthart**
HttpPayload::Read and HttpPayload::Trailers are tested by integration tests in a forthcoming PR.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083)

### CODE REVIEW

File: google/cloud/internal/http_payload.cc [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  return impl_->Read(buffer);
+}
+
+std::multimap<std::string, std::string> CurlHttpPayload::Trailers() const {
+  std::multimap<std::string, std::string> trailers;
+  auto headers = impl_->headers();
+  auto iter = std::find_if(headers.begin(), headers.end(),
+                           [](std::pair<std::string, std::string> p) {
+                             absl::AsciiStrToLower(&p.first);
+                             return p.first == "trailer";
```
Comment (dbolduc) [inline google/cloud/internal/http_payload.cc:41]: **not a full review, just a single comment:**

there is `absl::EqualsIgnoreCase()`. I'm guessing it's slightly faster than converting the whole string, then checking for equality. https://github.com/abseil/abseil-cpp/blob/b2c96417bd5c2b0a550611e503002a4594a932b2/absl/strings/match.h#L80-L81

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083)* status: all files reviewed, 9 unresolved discussions (waiting on @scotthart)

___
*[google/cloud/internal/http_payload.h, line 31 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuD9i2n6PAd8SD-n7lu:-MuD9i2n6PAd8SD-n7lv:bswjtu7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.h#L31)):*
> ```objc
> class CurlImpl;
> 
> class HttpPayload {
> ```

This class needs some documentation. What is it for? Who should be using it?
___
*[google/cloud/internal/http_payload.h, line 38 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuD9nsr9g0aACh8pkWk:-MuD9nsr9g0aACh8pkWl:bakb4xk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.h#L38)):*
> ```objc
> };
> 
> class CurlHttpPayload : public HttpPayload {
> ```

Consider moving this to a separate header, so folks don't need to `#include` all the curl stuff just to use an `HttpPayload`.
___
*[google/cloud/internal/http_payload.cc, line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDCccS89raCwQlwKs6:-MuDCccT-9MQGGNT6eVo:b-2y014p) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L37)):*
> ```cpp
> std::multimap<std::string, std::string> CurlHttpPayload::Trailers() const {
>   std::multimap<std::string, std::string> trailers;
>   auto headers = impl_->headers();
> ```

I don't recall how `impl_->headers()` works, but sounds to be that all this complexity should be in the `CurlImpl` class.  Otherwise there are two places that "know" how to the trailers are tunneled into the headers.
___
*[google/cloud/internal/http_payload.cc, line 40 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDAPlp2fwwxmzzwUT-:-MuDAPlp2fwwxmzzwUT0:b-czo8vm) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L40)):*
> ```cpp
>   auto iter = std::find_if(headers.begin(), headers.end(),
>                            [](std::pair<std::string, std::string> p) {
>                              absl::AsciiStrToLower(&p.first);
> ```

Sounds like all this code could be simplified if the keys were stored in lowercase?  Also, I do not know of a single case where GCP sends or receives trailers.  Just saying.
___
*[google/cloud/internal/http_payload.cc, line 63 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDAoDO3fDTILXqnZ82:-MuDAoDO3fDTILXqnZ83:b32fpeu) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L63)):*
> ```cpp
>   // Allocate buf on the heap as large values of read_size could exceed stack
>   // size.
>   auto buf = absl::make_unique<char[]>(read_size);
> ```

Why not `std::vector<char>`?  It is guaranteed to be contiguous.
___
*[google/cloud/internal/http_payload.cc, line 68 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDB2g89vz_d9Q55IlP:-MuDB2g89vz_d9Q55IlQ:bqiktnx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L68)):*
> ```cpp
>     read_status = payload->Read({&buf[0], read_size});
>     if (!read_status.ok()) return read_status.status();
>     output_buffer.write(buf.get(), read_status.value());
> ```

sounds like `output_buffer` should simply be a `std::string` and you need to do `output_buffer.append()`?
___
*[google/cloud/testing_util/CMakeLists.txt, line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDC-KEBBDSk4O1JUyb:-MuDC-KEBBDSk4O1JUyc:b-rfkrr1) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/testing_util/CMakeLists.txt#L37)):*
> ```CMake
>         mock_fake_clock.cc
>         mock_fake_clock.h
>         mock_rest_utils.h
> ```

This means `google_cloud_cpp_testing` now depends on `google_cloud_cpp_rest_internal` I would think we want to avoid that dependency?  It will force a relink of everything when the REST library changes.
___
*[google/cloud/testing_util/mock_rest_utils.h, line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDBhtm4D6ajVlZYZC0:-MuDBhtm4D6ajVlZYZC1:b-w43q0s) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/testing_util/mock_rest_utils.h#L1)):*
> ```objc
> // Copyright 2022 Google LLC
> ```

This file should be called `mock_http_payload.h`.  Let's not multiply the `*_utils.` files.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083)* status: 2 of 12 files reviewed, 9 unresolved discussions (waiting on @coryan and @dbolduc)

___
*[google/cloud/internal/http_payload.h, line 31 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuD9i2n6PAd8SD-n7lu:-MuDp2yP8cG5Y2S-tvUG:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.h#L31)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This class needs some documentation. What is it for? Who should be using it?
</blockquote></details>

Done.
___
*[google/cloud/internal/http_payload.h, line 38 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuD9nsr9g0aACh8pkWk:-MuDp3ki2HOaGWByQ1Nz:betn2g5) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.h#L38)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Consider moving this to a separate header, so folks don't need to `#include` all the curl stuff just to use an `HttpPayload`.
</blockquote></details>

All of the curl stuff is behind a pimpl class. No curl stuff is included in this header. Additionally, until we have an alternate back end, anybody using HttpPayload is going to be using CurlHttpPayload. The interface class exists primarily for mocking.
___
*[google/cloud/internal/http_payload.cc, line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDCccS89raCwQlwKs6:-MuDpYxCCivd_TDEpx2j:brcfqon) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L37)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I don't recall how `impl_->headers()` works, but sounds to be that all this complexity should be in the `CurlImpl` class.  Otherwise there are two places that "know" how to the trailers are tunneled into the headers.
</blockquote></details>

libcurl only provides a header callback. It makes no distinction between trailers and headers. The HTTP standard indicates that any expected trailers need to be "declared" in a "Trailer" header. I would prefer to leave the filtering as part of this class's responsibilities.
___
*[google/cloud/internal/http_payload.cc, line 40 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDAPlp2fwwxmzzwUT-:-MuDs1IW3BK31abyvM9X:boagllq) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L40)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Sounds like all this code could be simplified if the keys were stored in lowercase?  Also, I do not know of a single case where GCP sends or receives trailers.  Just saying.
</blockquote></details>

CurlImpl could be modified to always lower case the header keys. Issue #8089 created.
___
*[google/cloud/internal/http_payload.cc, line 41 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuD6TGQDiA1lAfxhAjF-r1-41:-MuDsweZ3eevTxvEiyZa:b-k4vt0g) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L41)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

**not a full review, just a single comment:**

there is `absl::EqualsIgnoreCase()`. I'm guessing it's slightly faster than converting the whole string, then checking for equality. https://github.com/abseil/abseil-cpp/blob/b2c96417bd5c2b0a550611e503002a4594a932b2/absl/strings/match.h#L80-L81
</blockquote></details>

This will be remedied as part of fixing #8089.
___
*[google/cloud/internal/http_payload.cc, line 63 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDAoDO3fDTILXqnZ82:-MuDtAuXDMzVO6tR1v5B:b-scvoh2) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L63)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Why not `std::vector<char>`?  It is guaranteed to be contiguous.
</blockquote></details>

This seemed straightforward and was one line instead of a declaration and a call to reserve.
___
*[google/cloud/internal/http_payload.cc, line 68 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDB2g89vz_d9Q55IlP:-MuDuiAkEhDfXxGsO8ci:b-pkc8ms) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L68)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

sounds like `output_buffer` should simply be a `std::string` and you need to do `output_buffer.append()`?
</blockquote></details>

Changed from stringstream to string
___
*[google/cloud/testing_util/CMakeLists.txt, line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDC-KEBBDSk4O1JUyb:-MuDunX80tRoBnq_J4Sk:bwiytqy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/testing_util/CMakeLists.txt#L37)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This means `google_cloud_cpp_testing` now depends on `google_cloud_cpp_rest_internal` I would think we want to avoid that dependency?  It will force a relink of everything when the REST library changes.
</blockquote></details>

Split the rest mocks into a separate library.
___
*[google/cloud/testing_util/mock_rest_utils.h, line 1 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDBhtm4D6ajVlZYZC0:-MuDusuM2B5WvBGfIuzs:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/testing_util/mock_rest_utils.h#L1)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This file should be called `mock_http_payload.h`.  Let's not multiply the `*_utils.` files.
</blockquote></details>

Done.

**Comment by coryan** [comment]
___
*[google/cloud/internal/http_payload.cc, line 63 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDAoDO3fDTILXqnZ82:-MuE-EdL3FYaKDjP5DWN:btc88j2) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L63)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

This seemed straightforward and was one line instead of a declaration and a call to reserve.
</blockquote></details>

`std::vector<char> foo(read_size);` ?

**Comment by coryan** [comment]
___
*[google/cloud/internal/http_payload.cc, line 67 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuE-OkI0Bhc65xq09mI:-MuE-OkI0Bhc65xq09mJ:b-5tfz2y) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L67)):*
> ```cpp
>   do {
>     read_status = payload->Read({&buf[0], read_size});
>     if (!read_status.ok()) return read_status.status();
> ```

nit: `return std::move(read_status).status();`

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 10 of 10 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083)* status: all files reviewed, 5 unresolved discussions (waiting on @dbolduc and @scotthart)

___
*[google/cloud/internal/http_payload.h, line 38 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuD9nsr9g0aACh8pkWk:-MuE01C-8dPa4hx6p7Jk:b-saglur) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.h#L38)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

All of the curl stuff is behind a pimpl class. No curl stuff is included in this header. Additionally, until we have an alternate back end, anybody using HttpPayload is going to be using CurlHttpPayload. The interface class exists primarily for mocking.
</blockquote></details>

Hmmm... 

https://github.com/googleapis/google-cloud-cpp/blob/2914630ba3c501d2888449d249af1fd1654eef00/google/cloud/internal/curl_impl.h#L18-L20

Looks like including `curl_impl.h` drags a few more dependencies.  I know it is 2022, and modules will happen "Real Soon Now[tm]", but compilation barriers are still useful.  :shrug: 
___
*[google/cloud/internal/http_payload.cc, line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDCccS89raCwQlwKs6:-MuE0SU283E2ttiNK3fF:b5mtqab) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L37)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

libcurl only provides a header callback. It makes no distinction between trailers and headers. The HTTP standard indicates that any expected trailers need to be "declared" in a "Trailer" header. I would prefer to leave the filtering as part of this class's responsibilities.
</blockquote></details>

TIL: the `trailer` header tells you what headers are trailers.  But that is not the only way?

https://curl.se/libcurl/c/CURLOPT_HEADERFUNCTION.html#:~:text=1)%20it%20comes%20after%20the%20response%2Dbody.%202)%20it%20comes%20after%20the%20final%20header%20line%20(CR%20LF)

That suggests to me that this code:

https://github.com/googleapis/google-cloud-cpp/blob/2914630ba3c501d2888449d249af1fd1654eef00/google/cloud/internal/curl_impl.cc#L337-L340

Could so something like:

```cc
std::size_t CurlImpl::HeaderCallback(char* contents, std::size_t size,
                                     std::size_t nitems) {
  if (after_payload_) {
      return CurlAppendHeaderData(received_trailers_, contents, size * nitems);
  }
  return CurlAppendHeaderData(received_headers_, contents, size * nitems);
}
```

And then you could simplify the code here?
___
*[google/cloud/internal/http_payload.cc, line 40 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDAPlp2fwwxmzzwUT-:-MuE3GPnE-135U8QXObP:bb7ssjn) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L40)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

CurlImpl could be modified to always lower case the header keys. Issue #8089 created.
</blockquote></details>

It seems they already are:

https://github.com/googleapis/google-cloud-cpp/blob/2914630ba3c501d2888449d249af1fd1654eef00/google/cloud/internal/curl_wrappers.cc#L230-L231

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083)* status: all files reviewed, 4 unresolved discussions (waiting on @coryan and @dbolduc)

___
*[google/cloud/internal/http_payload.cc, line 37 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDCccS89raCwQlwKs6:-MuIY14V1M0-AdCLsDcw:bss6ji) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L37)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

TIL: the `trailer` header tells you what headers are trailers.  But that is not the only way?

https://curl.se/libcurl/c/CURLOPT_HEADERFUNCTION.html#:~:text=1)%20it%20comes%20after%20the%20response%2Dbody.%202)%20it%20comes%20after%20the%20final%20header%20line%20(CR%20LF)

That suggests to me that this code:

https://github.com/googleapis/google-cloud-cpp/blob/2914630ba3c501d2888449d249af1fd1654eef00/google/cloud/internal/curl_impl.cc#L337-L340

Could so something like:

```cc
std::size_t CurlImpl::HeaderCallback(char* contents, std::size_t size,
                                     std::size_t nitems) {
  if (after_payload_) {
      return CurlAppendHeaderData(received_trailers_, contents, size * nitems);
  }
  return CurlAppendHeaderData(received_headers_, contents, size * nitems);
}
```

And then you could simplify the code here?
</blockquote></details>

Per our discussion, I've created issue #8107 to defer supporting trailers for now.
___
*[google/cloud/internal/http_payload.cc, line 40 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDAPlp2fwwxmzzwUT-:-MuIZArF3TmWsqZTFlEH:b-lose4a) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L40)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

It seems they already are:

https://github.com/googleapis/google-cloud-cpp/blob/2914630ba3c501d2888449d249af1fd1654eef00/google/cloud/internal/curl_wrappers.cc#L230-L231
</blockquote></details>

Removing trailers support for now.
___
*[google/cloud/internal/http_payload.cc, line 63 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuDAoDO3fDTILXqnZ82:-MuIYDyN2jmZiOSWxPIx:b-rqiiyq) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L63)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

`std::vector<char> foo(read_size);` ?
</blockquote></details>

I would have liked to use std::array here as it cannot be resized like a vector can. However, the size is dynamic so I opted for a c-style array on the heap.
___
*[google/cloud/internal/http_payload.cc, line 67 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-MuE-OkI0Bhc65xq09mI:-MuIYdUEEQCXOArJWlHg:bb6un1j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/327298eb2e4f6646f86fcac41e608e2f6e11c582/google/cloud/internal/http_payload.cc#L67)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: `return std::move(read_status).status();` 
</blockquote></details>

Done

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8083#-:-MuInJiS29fnrqHN9ss8:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @dbolduc)

**Pull request merged**